README.md
gxrandr
=======

which used to switching mode for primary monitor and second.


ChangeLog
ChangeLog:


gxrandr 2.3 ( 2012-06-12 )

	1. moved all the pictures to share/ dir.
	
	2. moved all desktop files to share/ dir.
	
	3. moved all the source files to src/ dir.
	
	4. fixed a bug as follwing description says:
		(gxrandr:7166): Gtk-CRITICAL **: IA__gtk_widget_set_size_request: assertion `GTK_IS_WIDGET (widget)' failed
		
	5. refactored all the code in gxrandr.c, much easier to matain.
	
	6. fixed a bug: when the user switch to other modes,
		the user unplugged the projecter.
		
	7. placed the startup desktop file in more places, easier to locate this tool.
	
	8. add a limit: only one process is allowed to be running simultaneously.
	
	9. a stronger Makefile introduced.
	
	10. merged all patches from ShiAnSheng & PengYan.


src/utility.c
#include <stdio.h>
#include <stdlib.h>

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>

#include <unistd.h>
#include <pwd.h>

#include "gxrandr.h"

char monitors[2][64]; // 显示器接口名
char mode_name[32]; // mode, eg. 1024x768 1399x768 ...
int valid_monitor_number = TRUE;
char cvt_string[128] = "";

static void read_cvt(char* buf) {
	FILE* fp = fopen("/tmp/gxrandr_cvt", "r");
	fgets(buf, 128, fp);
	fclose(fp);
}

char* cvt(int width, int height)
{
	char cmd[256];
	sprintf(cmd, "cvt %d %d | grep Modeline | awk -F' ' '{print $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13}' > /tmp/gxrandr_cvt", width, height);
	system(cmd);
	char* result = (char*)malloc(sizeof(char) * 128);
	read_cvt(result);
	system("rm -f /tmp/gxrandr_cvt");
	
	return result;
}

// ensures only one process is running at the same time
void single_process() {
	int pid_file = open("/tmp/gxrandr.pid", O_CREAT | O_RDWR, 0666);

	FILE* fp = fopen("/tmp/gxrandr.pid", "w");
	fprintf(fp, "%d\n", getpid());
	fclose(fp);

	int lock = lockf(pid_file, F_TLOCK, 0);
	
	if ( lock != 0 ) {
		fprintf(stderr, "gxrandr is already running\n");
		exit(1);
	}
}

void get_cvt_string()
{
	int width = 0;
	int height = 0;

	sscanf(mode_name, "%dx%d", &width, &height);
	
        width=1024;
        height=768;
        printf ("### width: %d\t height: %d\n", width, height);
	char* buf = cvt(width, height);
	strcpy(cvt_string, buf);
	free(buf);
}

// use xrandr to perform the real configuration change
void real_change_mode(int mode)
{
	char command[256] = "";
	char cmd[256] = "";
	char* cvt_str = NULL;
	int com;

	switch(mode) {
	    case COMPUTER_MODE:
	        sprintf(command, "gxrandr-switch-mode.sh COMPUTER");
		break;
	    case DUPLICATE_MODE:
	        sprintf(command, "gxrandr-switch-mode.sh DUPLICATE");
		break;
	    case EXTENSION_MODE:
	        sprintf(command, "gxrandr-switch-mode.sh EXTENSION");
		break;
	    case PROJECTER_MODE:
	        sprintf(command, "gxrandr-switch-mode.sh PROJECTOR");
		break;
	    default:
	        sprintf(command, "gxrandr-switch-mode.sh");
		break;
	}
	system(command);
	
	switch(mode) {
		case COMPUTER_MODE:
			sprintf(cmd, "echo %d > ~/.gxrandr", COMPUTER_MODE);
			break;
		case DUPLICATE_MODE:
			sprintf(cmd, "echo %d > ~/.gxrandr", DUPLICATE_MODE);
			break;
		case EXTENSION_MODE:
			sprintf(cmd, "echo %d > ~/.gxrandr", EXTENSION_MODE); 
			break;
		case PROJECTER_MODE:
			sprintf(cmd, "echo %d > ~/.gxrandr", PROJECTER_MODE); 
			break;
	    }
	    
	//printf("%s\n", command);
	//system(command);
	system("touch ~/.gxrandr");
	system(cmd);
}

// read home dir name to result.
static void get_home_dir(char* result) {
	uid_t user_id = getuid();
	struct passwd *pw = getpwuid(user_id);
	strcpy(result, pw->pw_dir);
}

static void get_monitor_names() {
	FILE* fp = NULL;
	int tag = system("test `xrandr | grep \' connected\' | wc -l` -eq 2");

	if ( tag != 0 ) {
		// 目前仅支持2个显示器
		valid_monitor_number = FALSE;
		strcpy(monitors[0], "LVDS");
		strcpy(monitors[1], "VGA-0");
	}
	else {
		system("xrandr | grep \' connected\' | awk -F' ' '{print $1}' > /tmp/gxrandr");

		fp = fopen("/tmp/gxrandr", "r");
		if ( fp == NULL ) {
			printf("[error] cannot read /tmp/gxrandr\n");
			strcpy(monitors[0], "LVDS");
			strcpy(monitors[1], "VGA-0");
		}
		else {
			fscanf(fp, "%s\n", monitors[0]);
			fscanf(fp, "%s\n", monitors[1]);
			fclose(fp);
		}

		// if LVDS is dectected as the second monitor, make it the first one.
		//if ( strcmp(monitors[1], "LVDS") == 0 ) {
		//	char str[128] = "";
		//	strcpy(str, monitors[0]);
		//	strcpy(monitors[0], "LVDS");
		//	strcpy(monitors[1], str);
		//}
		// if DP5 is dectected as the second monitor, make it the first one.
		if ( strcmp(monitors[1], "DP5") == 0 ) {
			char str[128] = "";
			strcpy(str, monitors[0]);
			strcpy(monitors[0], "DP5");
			strcpy(monitors[1], str);
		}
		if ( strcmp(monitors[1], "DP6") == 0 ) {
			char str[128] = "";
			strcpy(str, monitors[0]);
			strcpy(monitors[0], "DP6");
			strcpy(monitors[1], str);
		}
		if ( strcmp(monitors[1], "LCD") == 0 ) {
			char str[128] = "";
			strcpy(str, monitors[0]);
			strcpy(monitors[0], "LCD");
			strcpy(monitors[1], str);
		}
	}
}

// guess the latest projecter mode
int guess() {
	int mode = 0;
	FILE* fp = NULL;

	if ( system("test -f ~/.gxrandr") == 0 ) {
		char* home = (char*)alloca(sizeof(char) * 256);
		get_home_dir(home);
		strcat(home, "/.gxrandr");

		fp = fopen(home, "r");
		fscanf(fp, "%d", &mode);
		fclose(fp);
	}

	get_monitor_names();

	if ( system("test -f /tmp/gxrandr-mode") != 0 ) {
		system("xrandr | grep \" connected\" | awk -F' ' '{ print $4 }' | awk -F'+' '{print $1}' | grep -e \"[0-9]*x[0-9]*\" > /tmp/gxrandr-mode");
		if ( system("xrandr |grep \" connected\" |grep DP |grep -v primary") == 0 ){
			system("xrandr | grep \" connected\" |grep DP | awk -F' ' '{ print $3 }' | awk -F'+' '{print $1}' | grep -e \"[0-9]*x[0-9]*\" > /tmp/gxrandr-mode");
		}
	}
		if ( system("xrandr |grep \" connected\" |grep DP |grep primary ") == 0 ){
			system("xrandr | grep \" connected\" |grep DP | awk -F' ' '{ print $4 }' | awk -F'+' '{print $1}' | grep -e \"[0-9]*x[0-9]*\" > /tmp/gxrandr-mode");
		}
	fp = fopen("/tmp/gxrandr-mode", "r");
	fscanf(fp, "%s", mode_name);
	sprintf(mode_name,"%s","1024x768_60.00");
	fclose(fp);

	return mode;
}


check_out_screen.sh
#!/bin/bash

while [ 1 ]
do
	num_screen=`xrandr |grep connected |grep -v disconnected |wc -l`
	
	case "$num_screen" in
		'1' )
			#echo "1"
			size_mode=`xrandr | grep " connected" | awk -F' ' '{ print $4 }' | awk -F'+' '{print $1}' | grep -e "[0-9]*x[0-9]*"`
			in_screen=`xrandr |grep " connected"|awk '{print $1}'`
			if [ -f "/tmp/gxrandr-mode" ];then
				size_mode_exit=`cat /tmp/gxrandr-mode`
				if [ "$size_mode_exit" != "$size_mode" ];then
					echo $size_mode >/tmp/gxrandr-mode
				fi
			else
				echo $size_mode >/tmp/gxrandr-mode
			fi
			if [ ! -f ~/.in_screen ];then
				echo $in_screen >~/.in_screen
			fi
			if [ -f ~/.num_screen ];then
				num_screen_exit=`cat ~/.num_screen`
				if [ "$num_screen_exit" != "$num_screen" ];then
					echo $num_screen >~/.num_screen
				fi
			else
				echo $num_screen >~/.num_screen
			fi
			echo "0" >~/.gxrandr
		;;
		'2' )
			#echo "2"
			if [ -f ~/.num_screen ];then
				num_screen_exit=`cat ~/.num_screen`
				#echo "$num_screen_exit-----$num_screen"
				if [ "$num_screen_exit" != "$num_screen" ];then
					in_screen=`cat ~/.in_screen`
					size_mode=`cat /tmp/gxrandr-mode`
					out_screen=`xrandr |grep " connected" |grep -v "$in_screen" |awk '{print $1}'`
					wight=`echo $size_mode |awk -F 'x' '{print $1}'`
                                        hight=`echo $size_mode |awk -F 'x' '{print $2}'`
					if [ "$in_screen"  != "DP5" ];then
                                        	cvt_string=`cvt $wight $hight |grep -v \# |sed 's/Modeline//g'` 
                                        	new_size_mode=`echo $cvt_string |awk -F '"' '{print $2}'`
						cvt_string_2=`echo $cvt_string |awk -F '"' '{print $3}'`
                                        	xrandr --newmode $new_size_mode $cvt_string_2
                                        	xrandr --addmode $out_screen $new_size_mode

						xrandr  --output $in_screen --auto --mode $size_mode  --output $out_screen --off
						usleep 10000
						xrandr --output $in_screen --mode $size_mode --output $out_screen --mode $new_size_mode --same-as $in_screen
					else
						xrandr  --output $in_screen --auto --mode $size_mode  --output $out_screen --off
						usleep 10000
						xrandr --output $in_screen --auto --output $out_screen  --auto --same-as $in_screen
					fi
					echo "1" >~/.gxrandr
					echo $num_screen >~/.num_screen
				fi
			else	
				echo $num_screen >~/.num_screen
			fi
		;;
	esac
	sleep 1
done


src/project-daemon.sh
#!/bin/bash
# ellenpeng <yan.peng@cs2c.com.cn> @ 2012-05-30
# This script want to detect the VGA plug/unplug event.

# display_old, display_new is the number of display 
# to compare the display_old and display_new to check the plug/unplug event
display_old=1
/bin/cat /proc/cmdline |grep notebook  >/dev/null 2>&1
if [ 1 == $? ];then
   exit
fi

while true
do
	sleep 2
   # use drm drvier, just for opensource intel/ati/nv driver
	if [ -d /sys/class/drm/card0 ]; then
		display_new=0
		for i in `ls /sys/class/drm/card*/status`
		do
			if [ `cat $i` = "connected" ];then
				display_new=$((display_new+1))
			fi
		done
   fi

   if [ $display_old -ne $display_new ]; then
		if [ $display_new -eq 2 ];then
			# dual display, change the display to clone mode
			/usr/bin/gxrandr &
		fi
   fi
   
   if [ $display_new -eq 1 -a -f /tmp/gxrandr.pid ]; then
		if [ `cat ~/.gxrandr` -eq 3 ]; then
			pid=`cat /tmp/gxrandr.pid`
			kill -9 $pid
			rm -f /tmp/gxrandr.pid
			#/usr/bin/gxrandr -m 1
			xrandr --auto
		fi
   	fi
   display_old=$display_new
done


gxrandr-switch-mode.sh
#! /bin/bash

function add_newmode_to_monitor()
{
    monitor=$1
    modeline=$2
    resolution=$3

    echo "### monitor:${monitor}" >> ${LOG}
    echo "### modeline:${modeline}" >> ${LOG}
    echo "### resolution:${resolution}" >> ${LOG}

    echo "### cmd: xrandr --newmode ${modeline}" >> ${LOG}
    xrandr --newmode ${modeline} &>/dev/null
    xrandr --newmode `echo ${modeline} | sed -r 's/"//g'`
    echo "### cmd: xrandr --addmode ${monitor} ${resolution}" >> ${LOG}
    xrandr --addmode ${monitor} ${resolution} #&>/dev/null
}

function add_and_set_newmode_to_main_monitor()
{
    echo "### add_newmode_to_main_monitor ###" >> ${LOG}
    modeline="$(echo ${S_MAX_RES} | tr -s 'x' ' '| xargs cvt | grep "Modeline" |sed -r 's/Modeline//g' | sed -r 's/_60.00//g')"
    resolution="$(echo ${modeline} | awk '{print $1}' | sed -r 's/"//g')"

    ### 增加1024x768 分辨率到主屏。
    add_newmode_to_monitor ${M_MONITOR} "${modeline}" ${resolution}

    ### 设置复制模式
    echo "### cmd: xrandr --output ${S_MONITOR} --mode ${resolution} --auto --output ${M_MONITOR} --mode ${resolution} --same-as ${S_MONITOR}">>${LOG}
    xrandr --output ${S_MONITOR} --mode ${resolution} --auto --output ${M_MONITOR} --mode ${resolution} --same-as ${S_MONITOR}
}

function add_and_set_newmode_to_second_monitor()
{
    echo "### add_newmode_to_second_monitor ###" >> ${LOG}
    modeline="$(echo ${M_MAX_RES} | tr -s 'x' ' '| xargs cvt | grep "Modeline" |sed -r 's/Modeline//g' | sed -r 's/_60.00//g')"
    resolution="$(echo ${modeline} | awk '{print $1}' | sed -r 's/"//g')"

    ### 增加主屏的最大分辨率到副屏；
    add_newmode_to_monitor ${S_MONITOR} "${modeline}" ${resolution}
 
    ### 设置复制模式
    echo "### cmd: xrandr --output ${M_MONITOR} --mode ${M_MAX_RES} --auto --output ${S_MONITOR} --mode ${resolution} --same-as ${M_MONITOR}">> ${LOG}
    xrandr --output ${M_MONITOR} --mode ${M_MAX_RES} --auto --output ${S_MONITOR} --mode ${resolution} --same-as ${M_MONITOR}
}

function checking_resolution_max_between_main_and_second()
{
    echo "### M_MAX_RES: ${M_MAX_RES} # S_MAX_RES: ${S_MAX_RES}" >> ${LOG}
    M_RES_EQUATION=`echo ${M_MAX_RES} | sed -r 's/x/*/g'`
    M_RES_MUL=$[${M_RES_EQUATION}]
    S_RES_EQUATION=`echo ${S_MAX_RES} | sed -r 's/x/*/g'`
    S_RES_MUL=$[${S_RES_EQUATION}]
   
    if [ ${M_RES_MUL} -gt ${S_RES_MUL} ]
    then
        # main monitor add mode of second max resolution. 
        echo "### M_RES_MUL:${M_RES_MUL} greate than SRES_MUL:${S_RES_MUL}" >> ${LOG}
        return 1
    elif [ ${M_RES_MUL} -lt ${S_RES_MUL} ]
    then
        # second monitor add mode with max of main monitor "1366x768" or "1600x900"
        echo "### M_RES_MUL:${M_RES_MUL} little than S_RES_MUL:${S_RES_MUL}" >> ${LOG}
        return 2
    else
        # no operations.
        echo "### M_RES_MUL:${M_RES_MUL} equal S_RES_MUL:${S_RES_MUL}" >> ${LOG}
        return 0
    fi
}

function set_computer_mode()
{
    echo "### switching mode: computer" >> ${LOG}
    echo "### cmd: xrandr --output ${M_MONITOR} --auto --output ${S_MONITOR} --off" >> ${LOG}
    xrandr --output ${M_MONITOR} --auto --output ${S_MONITOR} --off
}

function set_duplicate_mode()
{
    echo "### switching mode: duplicate" >> ${LOG}
    checking_resolution_max_between_main_and_second; res=$?
    if [ ${res} -eq 1 ]; then
        echo "### main monitor have the max resolution, Main Monitor # ${res}" >> ${LOG}
        add_and_set_newmode_to_main_monitor
    elif [ ${res} -eq 2 ];then
        echo "### second monitor have the max resolution, Second Monitor # ${res}" >> ${LOG}
        add_and_set_newmode_to_second_monitor
    else
        echo "### max of main screen same as second screen." >> ${LOG}
    fi
}

function set_extension_mode()
{
    echo "### switching mode: extension " >> ${LOG}
    echo "### cmd:xrandr --output ${M_MONITOR} --auto --output ${S_MONITOR} --right-of ${M_MONITOR}" >> ${LOG}
    xrandr --output ${M_MONITOR}  --auto  --output ${S_MONITOR} --right-of ${M_MONITOR}
}

function set_projector_mode()
{
    echo "### switching mode: projector # " >> ${LOG}
    echo "### cmd:xrandr --output ${S_MONITOR} --auto --output ${M_MONITOR} --off" >> ${LOG}
    xrandr --output ${S_MONITOR} --auto --output ${M_MONITOR} --off
}

function check_and_set_mode()
{
    mode=$1
    if [ "${mode}" = "COMPUTER" ]
    then
        set_computer_mode
    elif [ "${mode}" = "DUPLICATE" ]
    then
        set_duplicate_mode
    elif [ "${mode}" = "EXTENSION" ]
    then
        set_extension_mode
    elif [ "${mode}" = "PROJECTOR" ]
    then
        set_projector_mode
    else
        echo "### switching mode: Unknown" >> ${LOG}
    fi
}

MODE="$1"; N_NUM=""; M_MONITOR=""; S_MONITOR=""; M_MAX_RES=""; S_MAX_RES=""
#TIME=`date "+%y%m%d%H%M%S"`
#LOG="/dev/stdout"
LOG="/tmp/xrandr.log"
if [ -e ${LOG} ];then rm ${LOG}; fi

### M_NUM:Monitor numbers. 判断是否双显示器
M_NUM="$(xrandr | grep "\<connected\>" | wc -l)" 
if [ "${M_NUM}" = "2" ];then
    xrandr | grep "\<connected\>" | grep "DP5"; res=$?
    echo "### res: $res"
    if [ ${res} -eq 0 ]
    then
        echo "### primary monitor is DP5" >> $LOG
        M_MONITOR="$(xrandr  | grep "\<connected\>" | sed -n '2 p' | awk '{print $1}')"
        S_MONITOR="$(xrandr | grep -m 1 "\<connected\>" | cut -d ' ' -f1)"
    else
        echo "### primary monitor is not DP5" >> $LOG
        M_MONITOR="$(xrandr | grep -m 1 "\<connected\>" | cut -d ' ' -f1)"
        S_MONITOR="$(xrandr  | grep "\<connected\>" | sed -n '2 p' | awk '{print $1}')"
    fi
    xrandr --output ${M_MONITOR} --auto --output ${S_MONITOR} --off
    M_MAX_RES="$(xrandr | grep "*" | awk '{print $1}' | sed -n '1 p')"
    xrandr --output ${S_MONITOR} --auto --output ${M_MONITOR} --off
    S_MAX_RES="$(xrandr | grep "*" | awk '{print $1}' | sed -n '1 p')"
    
    echo "### MonitorNum:${M_NUM} #M_MONITOR: ${M_MONITOR} #S_MONITOR: ${S_MONITOR} #M_MAX_RES: ${M_MAX_RES} #S_MAX_RES: ${S_MAX_RES}" >> ${LOG}
    ### 判断需要设置的模式
    check_and_set_mode ${MODE}
else
    echo "### ERROR, Please checking monitor number." >> ${LOG}
    exit -1
fi


src/gxrandr.c
/*******************************************************************
 * Written by: Wu Xiao
 * Date: 2011-12-12
 * Updated: 2012-02-06
 * UPdated: 2012-06-11
 ******************************************************************/

#include <errno.h>
#include <gtk/gtk.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>

#include "gxrandr.h"
#include "utility.h"

static GtkWidget *window = NULL; // 主窗口
static GtkWidget *fixed = NULL; // 固定框(不可见，容器)
GtkWidget *horizontal_box = NULL;  // 水平容器
static GtkWidget* frame; // 选择框
static GtkWidget* image[4]; // 图片
static GtkWidget* event_box[4];

static GtkWidget* closebtn = NULL;
static GtkWidget* closebtn_event_box = NULL;

static int frame_pos[4];
static int current_mode = -1;

int screen_width = 1024;
int screen_height = 768;
int win_width = 512;
int win_height = 128;

// forward declarations
static gint timeout_cb(gpointer data);
static void get_display_metrics();
static void update_widget_sizes();

static void change_mode(int mode) {
	if ( mode < 0 || mode > 3 ) {
		g_print("[ERROR] Unrecognizable mode(%d)\n",mode);
		return;
	}

#if 1
	if ( current_mode != mode ) {//xiaoying.hu set image of button 
		//gtk_widget_set_visible(frame[mode], TRUE);
		gtk_fixed_move(GTK_FIXED(fixed), frame, frame_pos[mode], 0);
		current_mode = mode;
	}
#endif
}

static gint timeout_tag = 0;
static void add_timeout(gint ms) {
	if ( timeout_tag != 0 ) {
		g_source_remove(timeout_tag);
	}
	timeout_tag = g_timeout_add(ms, timeout_cb, NULL);
}

static gint timeout_cb(gpointer data) {
	real_change_mode(current_mode);
	gtk_main_quit();
	return FALSE;
}

static void get_display_metrics()
{
	GdkScreen* screen = gdk_screen_get_default();
	screen_width = gdk_screen_get_width(screen);
	screen_height = gdk_screen_get_height(screen);
	//g_print("screen.width = %d\tscreen.height = %d\n", screen_width, screen_height);
	
	if ( screen_width > 1440 ) {
		screen_width = 1440;
	}

	win_width = screen_width / 2;
	win_height = win_width / 4;

	//win_width += 32;

	int i = 0;
	for ( i = 0; i < 4; i++ ) {
		frame_pos[i] = i * win_height;
	}
}

static void update_widget_sizes()
{
	get_display_metrics();
	//gtk_widget_set_size_request(window, win_width, win_height);
	gtk_window_resize(GTK_WINDOW(window), win_width, win_height);
	gtk_widget_set_size_request(
		frame,
		win_height,
		win_height
	);

	int i = 0;
	for ( i = 0; i < 4; i++ ) {
		gtk_widget_set_size_request(
			GTK_WIDGET(image[i]),
			win_height,
			win_height
		);
	}
}


static gboolean press_handler(GtkWidget* widget, GdkEvent* event, gpointer user_data)
{
	//g_print("arg = %d\n", (int)user_data);
	char command[256] = "";
	change_mode((int)user_data);
        //sprintf(command, "gxrandr-switch-mode.sh ", );
        //system(command);

	add_timeout(10);
	return FALSE;
}

static gboolean close_handler(GtkWidget* widget, GdkEvent* event, gpointer user)
{
	gtk_main_quit();
	system("rm -f /tmp/gxrandr.pid; rm -f /tmp/gxrandr");
	return FALSE;
}

static GdkPixbuf* closebtn_pixbuf[2];
// 鼠标进入关闭按钮时被调用
static gboolean closebtn_enter_handler(
	GtkWidget* widget,
	GdkEvent* event,
	gpointer user)
{
	gtk_image_set_from_pixbuf((GtkImage*)closebtn, closebtn_pixbuf[1]);
	return FALSE;
}

// 鼠标离开关闭按钮时被调用
static gboolean closebtn_leave_handler(
	GtkWidget* widget,
	GdkEvent* event,
	gpointer user)
{
	gtk_image_set_from_pixbuf((GtkImage*)closebtn, closebtn_pixbuf[0]);
	return FALSE;
}

// 加载图片，放大缩小，再传给新的GtkImage
GtkWidget* gtk_image_new_from_file_at_scale(
	const gchar *filename,
	int width,
	int height)
{

	GdkPixbuf* pixbuf = NULL;

	pixbuf = gdk_pixbuf_new_from_file_at_scale(
		filename, width, height, FALSE, NULL);
	if ( pixbuf == NULL ) {
		return NULL;
	}

	GtkWidget* image = gtk_image_new_from_pixbuf(pixbuf);
	return image;
}

static int cmd_mode = 0;
static int cmd_mode_no = 0;

void init(int argc, char* argv[]) {

	// getopt, getopt_long, getopt_long_only, optarg, opterr, optopt
	int c = getopt(argc, argv, "m:");
	
	if ( c == 'm' ) {
		cmd_mode = 1;
		cmd_mode_no = atoi(optarg);
		real_change_mode(cmd_mode_no - 1);
		exit(0);
	}
	else {
		single_process();
		gtk_init(&argc, &argv);
		get_display_metrics();
	}
}

void create_main_window() {

	window = gtk_window_new(GTK_WINDOW_POPUP);
	//gtk_window_set_resizable(GTK_WINDOW(window), TRUE);
	//gtk_widget_set_size_request(window, win_width , win_height);
	gtk_window_resize(GTK_WINDOW(window), win_width, win_height);
	gtk_window_set_opacity(GTK_WINDOW(window), 0.8f);

	gtk_window_set_position(
		GTK_WINDOW(window),
		GTK_WIN_POS_CENTER_ALWAYS
	);

	//destroy-event
	g_signal_connect( G_OBJECT(window), "destroy",
		G_CALLBACK(gtk_main_quit), NULL);

	gtk_container_set_border_width(GTK_CONTAINER(window), 0);
}


void create_close_button() {
	// close button
	closebtn_pixbuf[0] = gdk_pixbuf_new_from_file_at_scale(
		CLOSEBTN_DEFAULT_PNG, 32, 32, FALSE, NULL);

	closebtn_pixbuf[1] = gdk_pixbuf_new_from_file_at_scale(
		CLOSEBTN_ENTERED_PNG, 32, 32, FALSE, NULL);

	closebtn = gtk_image_new_from_pixbuf(closebtn_pixbuf[0]);
/*	closebtn = gtk_image_new_from_file_at_scale(CLOSEBTN_PNG, 32, 32);*/
	
/*	closebtn = gtk_image_new_from_file(CLOSEBTN_PNG);*/
/*	gtk_widget_set_size_request(closebtn, 32, 32);*/
	closebtn_event_box = gtk_event_box_new();
	gtk_container_add(GTK_CONTAINER(closebtn_event_box), closebtn);

	gtk_fixed_put(GTK_FIXED(fixed), closebtn_event_box, win_width, 0);
	gtk_widget_show(closebtn_event_box);
	gtk_widget_show(closebtn);
	gtk_event_box_set_visible_window((GtkEventBox*)closebtn_event_box, FALSE);
	
	g_signal_connect(G_OBJECT(closebtn_event_box), "button-press-event", G_CALLBACK(close_handler), NULL);
	g_signal_connect(G_OBJECT(closebtn_event_box), "enter-notify-event", G_CALLBACK(closebtn_enter_handler), NULL);
	g_signal_connect(G_OBJECT(closebtn_event_box), "leave-notify-event", G_CALLBACK(closebtn_leave_handler), NULL);
}

void create_fixed_widget() {
	// fixed
	fixed = gtk_fixed_new();
	gtk_container_add(GTK_CONTAINER(window), fixed);
	gtk_widget_show(fixed);
	gtk_container_set_border_width(GTK_CONTAINER(fixed), 0);
}

void create_window_background() {
	GtkWidget* background = gtk_image_new_from_file_at_scale(
		BACKGROUND_PNG, win_width + 32, win_height);
	gtk_fixed_put(GTK_FIXED(fixed), background, 0, 0);
	gtk_widget_show(background);
}

void create_highlight_widget() {
	// 高亮窗口
	frame = gtk_image_new_from_file_at_scale(
		FRAME_PNG, win_height, win_height);
	gtk_fixed_put(GTK_FIXED(fixed), frame, 0, 0);
	gtk_widget_show(frame);
}

void create_horizontal_box() {
	// icon container - horizontal_box
	horizontal_box = gtk_hbox_new(TRUE, 0);
	gtk_container_set_border_width(GTK_CONTAINER(horizontal_box), 0);
	//gtk_container_add( GTK_CONTAINER(window), horizontal_box);
	gtk_container_add(GTK_CONTAINER(fixed), horizontal_box);
	gtk_widget_show(horizontal_box);
}

void create_main_buttons() {
	char* paths[4] = {
		COMPUTER_PNG,
		DUPLICATE_PNG,
		EXTENSION_PNG,
		PROJECTER_PNG
	};
	
	int projecter_mode[4] = {
		COMPUTER_MODE,
		DUPLICATE_MODE,
		EXTENSION_MODE,
		PROJECTER_MODE
	};
	
	int i = 0;
	for ( i = 0; i < 4; i++ ) {
		image[i] = gtk_image_new_from_file_at_scale(
				paths[i], win_height * 2 / 3, win_height * 2 / 3);
		gtk_widget_set_size_request(
			image[i], win_height, win_height
		);
	
		event_box[i] = gtk_event_box_new();
		gtk_container_add(GTK_CONTAINER(event_box[i]), image[i]);

		gtk_container_add(GTK_CONTAINER(horizontal_box), event_box[i]);
		gtk_widget_show(image[i]);
		gtk_widget_show(event_box[i]);
		gtk_event_box_set_visible_window((GtkEventBox*)event_box[i], FALSE);

		g_signal_connect(G_OBJECT(event_box[i]), "button-press-event", G_CALLBACK(press_handler), (gpointer)projecter_mode[i]);
	}
}

void create_user_interface() {

	create_main_window();
	create_fixed_widget();
	create_window_background();
	//create_highlight_widget();
	create_close_button();
	create_horizontal_box();
	create_main_buttons();
}

int main(int argc, char* argv[])
{
	int projecter_mode = 0;
	
	init(argc, argv);
	
	create_user_interface();
	//projecter_mode = guess();
	//change_mode(projecter_mode);

	gtk_widget_show(window);
	//add_timeout(3000);
	gtk_main();

	return 0;
}
